src/plugins/node/prof/watchdog.h

```diff
@@ -0,0 +1,330 @@
+  watchdog_activation_count.Exchange(0);
+  C::CpuProfiler::StartCpuProfiling(isolate);
+  // Block again in case V8 adds a pthread_sigmask(SIG_UNBLOCK).
+  CHECK_EQ(0, ::pthread_sigmask(SIG_BLOCK, &set, NULL));
+  CHECK(profiler_tid = FindProfilerTid());
+  CHECK_EQ(0, ::syscall(SYS_tgkill, ::getpid(), profiler_tid, kSuspendSignal));
+  // Arm timer that unblocks the profiler thread on expiry.
+  sigevent ev;
+  // Can't use ev.sigev_notify_thread_id because of broken glibc headers.
+  ev._sigev_un._tid = profiler_tid;
+  ev.sigev_notify = SIGEV_THREAD_ID;
+  ev.sigev_signo = kResumeSignal;
+  CHECK_EQ(0, ::timer_create(CLOCK_MONOTONIC, &ev, &timer_id));
+  CHECK_EQ(0, timeout.it_interval.tv_sec);
+  CHECK_EQ(0, timeout.it_interval.tv_nsec);
+  timeout.it_value.tv_sec = timeout_in_ms / 1000;
+  timeout.it_value.tv_nsec = (timeout_in_ms % 1000) * 1000 * 1000;
+  CHECK_EQ(0, ::timer_settime(timer_id, 0, &timeout, NULL));
+  CHECK_EQ(0, ::pthread_sigmask(SIG_UNBLOCK, &set, NULL));
+  return NULL;
+}
+
+const v8::CpuProfile* StopCpuProfiling(v8::Isolate* isolate) {
+  if (profiler_tid > 0) {
+    // Unblock profiler thread, V8 is about to pthread_join() it.
+    CHECK_EQ(0, ::syscall(SYS_tgkill, ::getpid(), profiler_tid, kResumeSignal));
+    CHECK_EQ(0, ::timer_settime(timer_id, 0, &no_timeout, NULL));
+    CHECK_EQ(0, ::timer_delete(timer_id));
+    profiler_tid = 0;
+  }
+  return C::CpuProfiler::StopCpuProfiling(isolate);
+}
+
+C::ReturnType WatchdogActivationCount(const C::ArgumentType& args) {
+  C::ReturnableHandleScope handle_scope(args);
+  return handle_scope.Return(watchdog_activation_count.Exchange(0));
+}
+
+void Initialize(v8::Isolate* isolate, v8::Local<v8::Object> binding) {
+  sigset_t mask;
+  ::sigemptyset(&mask);
+  ::sigaddset(&mask, kResumeSignal);
+  ::sigaddset(&mask, kSuspendSignal);
+  struct sigaction act;
+  ::memset(&act, 0, sizeof(act));
+  act.sa_handler = OnSignal;
+  act.sa_mask = mask;
+  CHECK_EQ(0, ::sigaction(kResumeSignal, &act, NULL));
+  CHECK_EQ(0, ::sigaction(kSuspendSignal, &act, NULL));
+  // See https://github.com/joyent/libuv/issues/1317.  We need to know when
+  // the program enters epoll_wait() but a bug in libuv makes prepare handles
+  // unsuitable for that because they fire at the wrong time.  That leaves us
+  // with little recourse but to hook the syscall() wrapper.
+  PatchSyscall();
+
+  v8::Local<v8::FunctionTemplate> watchdog_activation_count_template =
+      C::FunctionTemplate::New(isolate, WatchdogActivationCount);
+  binding->Set(
+      C::String::NewFromUtf8(isolate, "watchdogActivationCount"),
+      watchdog_activation_count_template->GetFunction());
+}
+
+}  // namespace watchdog
+
+#else  // defined(__linux__) && (defined(__i386) || defined(__x86_64__))
+
+namespace watchdog {
+
+namespace C = ::compat;
+
+void Initialize(v8::Isolate*, v8::Local<v8::Object>) {}
+
+const char* StartCpuProfiling(v8::Isolate* isolate, uint64_t timeout) {
```

Comment:
We should change this so that it is also possible at runtime to see if watchdog is supported without trying to do a start/stop to see if it failed.